### ISSUE TITLE
Review roles privileges

### ISSUE THREAD

**Issue by strk** [linz/linz-bde-schema#62]
I've been hit by errors of "permission denied" when even just creating a temporary schema during a BDE upload. This was due to some functions set as SECURITY DEFINER and being owned by the `bde_dba` role. I guess this may be a new error which came out since we stopped granting `bde_dba` the SUPERUSER role, which is something new in 1.1.0.

This ticket is to review the grants so to grant specific permissions to role that need them.

In other words, every time a function is set to be owned by `bde_dba` and made a `SECURITY DEFINER` we should make sure every action performed by that function (and functions called by that function) are permitted to the `bde_dba` user.

Not an easy task when security definers call other functions (more of a problem in linz-lds-bde-schema)

**Comment by strk** [comment]
In this repository only the `bde_drop_idle_connections` function is SECURITY DEFINER.
It is likely that such privilege is needed for calling `pg_terminate_backend`.
I'm not sure if RDS allows to call such function, btw

\cc @imincik

**Comment by imincik** [comment]
OK, looks like this problem is related: https://travis-ci.com/linz/bde-processor-deployment/jobs/101467396#L2763

**Comment by strk** [comment]
A good start here would be describing the intended use of created roles.
Such description could be encoded with `COMMENT ON ROLE` in `sql/01-bde_roles.sql`

**Comment by strk** [comment]
I've started a wiki page to help with gathering information:
https://github.com/linz/linz-bde-schema/wiki/Roles

**Comment by strk** [comment]
from looking at the gathered information I've the impression that removing SUPERUSER from `bde_dba` resulted in a drop of privileges from that user because it is never expliclty given permission (for example) on table revisions, while it is the owner of the original tables so I don't know if table_version functions themselves take care of that part.

**Comment by palmerj** [comment]
table-version tables get the same ownership as the source table. see https://github.com/linz/postgresql-tableversion/blob/master/sql/01-enable_versioning.sql#L76

**Comment by palmerj** [comment]
The role ownership comes from some long history of BDE use at LINZ. Roles were:

* dba- owns all objects (schema, tables, functions). Had rights to manage all of those things
* user - has rights to execute and read anything that dba creates. Used by users logging in to run analytics queries or for remote users get access to data (e.g LDS import or QGIS desktop users)
* admin - everything that user can do plus the ability to edit data in the tables. It was once thought that use cases needed a special user who was no dba to come and fix some data issues that the script didn't do. I think this role really can go now.

**Comment by strk** [comment]
Ok so roles are now documented. I'm closing this out for now. The `pg_terminate_backend` call may still be a problem on RDS...
